### PR TITLE
Add 5 Channel RL Train Script, One With Live CFD and Add Gaden Maps as Test on Eval

### PR DESCRIPTION
**Pull request by efemantarci**
## Overview

Two related changes ride on this branch:

1. **Channel-split** — `reinforcement_learning/` (old 4-channel module) is preserved untouched. Two parallel forks are added: `rl_5_channel/` (my 5-channel work) and `rl_cfd/` (my 6-channel CFD-extended work). Three coexisting modules instead of one breaking change.

2. **GADEN-maps evaluation** — a new evaluation pipeline that runs trained agents on the hand-curated maps in `base/gaden_maps/` (the same ones we use in the GADEN simulator), with their real source positions and CFD wind fields. Pipeline is wired into all three modules; env-side changes are backwards-compatible.

---

### Why three modules?

| Folder | Channels | Source | Status before this PR |
|---|---|---|---|
| `reinforcement_learning/` | 4 | friend (sonwezali) | already on `main`, training works |
| `rl_5_channel/` | 5 | mine | working-tree only, would break friend's checkpoints if merged in-place |
| `rl_cfd/` | 6 | mine (CFD extension) | working-tree only, divergent enough to need its own home |

Merging the 5-channel changes into `reinforcement_learning/` would break the friend's models. Keeping three modules side-by-side lets each evolve independently and lets old checkpoints keep loading.

### What's different across the three

**Observation channels:**
- 4: `[occupancy {-1 wall, 0 unknown, +1 free}, gas, recency, det_count]`
- 5: `[is_known, is_wall, gas, recency, det_count]` — splits occupancy into two binary channels so the static stream's first conv doesn't mix wall-edges with free-area gradients
- 6: `+ motion` — adds a recent-trajectory channel for the CFD module

**Policy ctx vector:**
- 4-ch: `(speed/max, dir/2π)` (wrap-around discontinuity at 0 / 2π)
- 5/6-ch: `(speed/max, cos(dir), sin(dir), step/MAX_STEPS)` — continuous direction encoding plus step progress

**rl_5_channel architectural fixes:**
- Reveal step uses the ground-truth grid + per-cell visibility check instead of LiDAR ray-casting (fixes a quantisation-overwrite bug where ray hits at 0.1 m wrote into 0.5 m wrapper cells, see [SPATIAL_CHANGES.md](base/rl_5_channel/SPATIAL_CHANGES.md))
- FiLM final layer initialised so γ=1, β=0 — starts as identity, learns a delta
- State-independent `actor_log_std` parameter (standard PPO recipe) instead of a state-conditional log-std head that saturates
- Optional batched-GPU obs path: thin worker procs send scalar state, main process batches reveal + ego embedding on GPU

**rl_cfd architectural additions:**
- New [`envs/wind_field.py`](base/rl_cfd/envs/wind_field.py) — spatially-varying wind from a 2D potential-flow solve (Dirichlet outer ring + Neumann interior obstacles) with optional divergence-free curl-noise overlay
- Plume queries `wind_field.local_batch(positions)` per filament instead of a uniform wind vector
- [`models/convert_5ch_to_6ch.py`](base/rl_cfd/models/convert_5ch_to_6ch.py) — checkpoint upconversion utility that reshapes `dynamic_cnn.0.weight` from `(16, 3, 5, 5)` → `(16, 4, 5, 5)`, zero-init the new motion-channel slot. Forward pass at fine-tune step 0 is bit-identical to the 5-channel checkpoint

---

### Why

The procedurally-generated `test_envs.json` evaluates against random rooms. The GADEN simulator runs against curated maps in `base/gaden_maps/` — and that's where the deployed agent lives. We need an eval that uses **the same walls, source positions, and wind fields** as deployment.

### How it works

**Geometry — STL rasterization:**

Each map has STL CAD files (`walls.stl`, `inner.stl`, `doors.stl`). We slice them at z=0.5 m (matches GADEN sim source height) and scanline-fill. GADEN's STL convention turned out to be:
- `walls.stl` → solid wall material (fill = wall) — outer perimeter
- `inner.stl` → fluid envelope (fill = free, complement = wall) — defines navigable region by what's inside it
- `doors.stl` → openings; ignored

So the final occupancy is `walls ∪ ¬inner`. Confirmed by visualizing each STL's fill independently. Wall fractions land at 10–20 % across the seven maps, matching real architecture.

No new dependency: binary STL parsed via `numpy.frombuffer`, scanline fill in pure numpy.

**Wind — CFD point-cloud rasterization with EDT gap fill:**

`wind_at_cell_centers_0.csv` is GADEN's OpenFOAM CFD output: 65k–670k `(Points:0/1/2, U:0/1/2)` rows, mesh density variable (0.0002–0.13 m nearest-neighbor distance). We:
1. Project all rows onto a regular 2D grid at `cell_size` (0.1 m) and average `(Ux, Uy)` per cell across z-layers
2. Many free cells get no CFD point because the mesh is sparser than 0.1 m in open regions — so we fill empty cells via `scipy.ndimage.distance_transform_edt(return_indices=True)`, copying wind from the nearest cell that does have data
3. Zero wind in walls

Coverage went from 40–65 % → **100 %** on every map after the EDT fill. The plume queries the field per filament with nearest-cell lookup; the policy ctx vector still gets a single `(speed, cos, sin, step)` derived from the field's spatial mean.

**Env injection — `options['map_data']`:**

`GasSourceEnv.reset(options={"map_data": ..., "wind_field": ...})` bypasses `MapGenerator.generate()` and uses the GADEN data directly. `WindModel.set_uniform()` (rl_5_channel + reinforcement_learning) pins the procedural model to the field's spatial mean for the policy ctx; `FilamentPlume` gains an optional `wind_field` kwarg and queries it per filament when set.

For `rl_cfd`, `GadenWindField` is shaped as a drop-in for the native `WindField` (`local_batch()`, `Ux/Uy` attrs, `get_observation_spatial()`, `get_dispersion_offset()`, `max_speed` ctx-norm attr). Reset swaps `self._wind` for the GADEN field for the episode and restores the procedural `WindField` on the next procedural reset.

**All env-side changes are backwards-compatible** — passing no options leaves the procedural training/eval path unchanged. Verified by re-running `eval_checkpoints_spatial` on the existing 100-env test set: results are bit-identical to before.

### New entry points (per module)

```bash
# Sanity-check the rasterized maps (writes test/gaden_viz/<map>.png)
python -m base.<module>.test.visualize_gaden_envs

# Visualize a checkpoint's trajectories (writes test/gaden_trajectories/<run>/<map>.png)
python -m base.<module>.test.visualize_gaden_trajectories \
    --run-dir base/<module>/runs/<run-dir> --episodes-per-map 1

# Full eval (per-map success / return per checkpoint)
sbatch base/<module>/eval_gaden.sh base/<module>/runs/<run-dir>
```

Maps used by default: `4rooms`, `uleft`, `uright`, `labyrinth_left`, `labyrinth_right`, `many_rooms`, `ultimate`. `ultimate`'s `sim1.robot_start` was changed from `(9, 3)` to `(19, 9)` to put the start on the opposite side of the building from the source.

### Verified results

End-to-end smoke test on the latest 5-channel checkpoint (`ppo_spatial_ent_20260424_164042_job4475`, step 99,975,168):

| Map | Success | Steps | Return |
|---|:-:|---:|---:|
| 4rooms | ✓ | 77 | +209 |
| uleft | ✓ | 31 | +206 |
| uright | ✓ | 49 | +208 |
| labyrinth_left | ✗ | 601 | -5957 |
| labyrinth_right | ✗ | 601 | -3241 |
| many_rooms | ✗ | 601 | -119 |
| ultimate (start 19,9) | ✗ | 601 | -1250 |

The labyrinth/many_rooms/ultimate failures are real — agent oscillates in tight corridors or wrong rooms. Trajectory PNGs make this immediately visible. The 6-channel checkpoint (`ppo_cfd_20260426_203843_job4497`) was also smoke-tested end-to-end on all 7 maps.

---

### Channel split (commits 1–4)

- `chore: gitignore experiment outputs and container images` — `runs/`, `results/`, `eval_results/`, `obs_viz/`, `*.sqsh`, `slurm-*.out`
- `feat(rrt-infotaxis): timestamp result paths and logs per run` — both `updated_weighted_rrt_infotaxis_igdm_*_discrete.py` files: per-run output dirs, log files, and result PNG suffixed with `YYYYMMDD_HHMMSS`
- `feat(rl_5_channel): add 5-channel spatial RL module` — 45 files (envs, models, training, test, viz scripts, READMEs, design docs, SLURM wrappers)
- `feat(rl_cfd): add 6-channel CFD-extended spatial RL module` — 39 files (same shape plus `wind_field.py`, `convert_5ch_to_6ch.py`, motion-channel viz)

### GADEN eval (commits 5–10)

- `chore: gitignore GADEN test output dirs` — `gaden_viz/`, `gaden_trajectories/`, `eval_results_gaden/`
- `feat(gaden_maps): set ultimate sim1 robot_start to (19, 9)` — yaml change
- `feat(rl_5_channel): GADEN map evaluation infrastructure` — 5 new files in `test/`, env modifications (`set_uniform`, `wind_field` kwarg, `options[map_data]`), `eval_gaden.sh`
- `feat(reinforcement_learning): GADEN map evaluation infrastructure` — same shape, adapted to 4-channel obs and 2-d wind ctx
- `feat(rl_cfd): GADEN map evaluation infrastructure` — `GadenWindField` extended with rl_cfd-compatible interface, reset() swaps wind for the episode
- `docs: document GADEN-maps eval in rl_5_channel and rl_cfd READMEs`

---

## Test plan

- [x] `python -m base.rl_5_channel.test.visualize_gaden_envs` produces 7 PNGs with sensible walls + wind quiver
- [x] `python -m base.rl_5_channel.test.visualize_gaden_trajectories --run-dir <run> --episodes-per-map 1` produces 7 trajectory PNGs
- [x] `sbatch base/rl_5_channel/eval_gaden.sh <run-dir>` writes per-map metrics to `test/eval_results_gaden/<run>.npz`
- [x] Same three commands work for `rl_cfd` against a 6-channel checkpoint
- [x] Existing `eval_spatial.sh` in `rl_5_channel` and `rl_cfd` still produces unchanged results (procedural path)
- [x] `python -m base.reinforcement_learning.test.visualize_gaden_envs` works (no 4-ch checkpoint to test trajectory viz against, but the loader and env modifications are exercised end-to-end with a random policy)
- [x] Procedural `GasSourceEnv.reset()` (no `options`) is bit-identical to before in all three modules

## Out of scope / future work

- Replaying GADEN's actual gas concentration field (`Sim_State_*.txt`) — currently we still seed a procedural FilamentPlume from the GADEN source position
- 4-channel checkpoint training in `reinforcement_learning/` to actually exercise the 4-ch GADEN trajectory viz
- Multi-sim coverage (`sim2`/`sim3` per map) — gated on filling in null `robot_start` values in the yaml
- Honouring `start_time` from the yaml — would matter if we later replay GADEN's plume

### CODE REVIEW

**Review by sonwezali — APPROVED** [review]
works properly. thanks for the contribution.
